storage/src/path/component.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+#[cfg(not(feature = "branch_factor_256"))]
+/// A path component in a hexary trie; which is only 4 bits (aka a nibble).
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PathComponent(pub crate::u4::U4);
+
+#[cfg(feature = "branch_factor_256")]
+/// A path component in a 256-ary trie; which is 8 bits (aka a byte).
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PathComponent(pub u8);
+
+impl PathComponent {
+    /// All possible path components.
+    ///
+    /// This makes it easy to iterate over all possible children of a branch
+    /// in a type-safe way. It is preferrable to do:
+    ///
+    /// ```ignore
+    /// for (idx, slot) in PathComponent::ALL.into_iter().zip(branch.children.each_ref()) {
+    ///     /// use idx and slot
+    /// }
+    /// ```
+    ///
+    /// instead of using a raw range like (`0..16`) or  [`Iterator::enumerate`],
+    /// which does not give a type-safe path component.
+    #[cfg(not(feature = "branch_factor_256"))]
+    pub const ALL: [Self; 16] = [
+        Self(crate::u4::U4::new_masked(0x0)),
+        Self(crate::u4::U4::new_masked(0x1)),
+        Self(crate::u4::U4::new_masked(0x2)),
+        Self(crate::u4::U4::new_masked(0x3)),
+        Self(crate::u4::U4::new_masked(0x4)),
+        Self(crate::u4::U4::new_masked(0x5)),
+        Self(crate::u4::U4::new_masked(0x6)),
+        Self(crate::u4::U4::new_masked(0x7)),
+        Self(crate::u4::U4::new_masked(0x8)),
+        Self(crate::u4::U4::new_masked(0x9)),
+        Self(crate::u4::U4::new_masked(0xA)),
+        Self(crate::u4::U4::new_masked(0xB)),
+        Self(crate::u4::U4::new_masked(0xC)),
+        Self(crate::u4::U4::new_masked(0xD)),
+        Self(crate::u4::U4::new_masked(0xE)),
+        Self(crate::u4::U4::new_masked(0xF)),
+    ];
+
+    /// All possible path components.
+    ///
+    /// This makes it easy to iterate over all possible children of a branch
+    /// in a type-safe way. It is preferrable to do:
+    ///
+    /// ```ignore
+    /// for (idx, slot) in PathComponent::ALL.into_iter().zip(branch.children.each_ref()) {
+    ///     /// use idx and slot
+    /// }
+    /// ```
+    ///
+    /// instead of using a raw range like (`0..256`) or  [`Iterator::enumerate`],
+    /// which does not give a type-safe path component.
+    #[cfg(feature = "branch_factor_256")]
+    pub const ALL: [Self; 256] = {
+        let mut all = [Self(0); 256];
+        let mut i = 0;
+        #[expect(clippy::indexing_slicing)]
+        while i < 256 {
+            all[i] = Self(i as u8);
+            i += 1;
+        }
+        all
+    };
```

Comment:
fortunately, `core::array::from_fn` is const as of rust 1.63, so this can become:
```suggestion
pub const ALL: [Self; 256] = core::array::from_fn(|i| Self(i as u8));
```